node/src/store.rs

```diff
@@ -33,6 +34,33 @@ pub trait Store: Send + Sync + Debug {
     /// Returns the header of a specific height.
     async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader>;
 
+    /// Returns the headers from the given heights range.
+    ///
+    /// If start of the range is unbounded, the first returned header will be of height 1.
+    /// height.
+    /// If end of the range is unbounded, the last returned header will be the last header in the
+    /// store.
+    ///
+    /// # Errors
+    ///
+    /// If range contains a height of a header that is not found in the store or [`RangeBounds`]
+    /// cannot be converted to a valid range.
+    async fn get_range<R>(&self, range: R) -> Result<Vec<ExtendedHeader>>
+    where
+        R: RangeBounds<u64> + Send,
+    {
+        let head_height = self.head_height().await?;
+        let range = to_headers_range(range, head_height)?;
+        let mut headers = Vec::with_capacity(range.clone().count());
```

Comment:
Is `count` optimized in this case? Or does it actually iterate all the items in the range?

Reply:
no, and it's even worse, it can [overflow on wasm](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#overflow-behavior) :exploding_head: 

I went through the possible options and found out about [`TrustedLen`](https://doc.rust-lang.org/stable/std/ops/struct.RangeInclusive.html#impl-TrustedLen-for-RangeInclusive%3CA%3E) and about [ExactSizeIterator](https://doc.rust-lang.org/stable/std/ops/struct.RangeInclusive.html#impl-ExactSizeIterator-for-RangeInclusive%3Cu16%3E) however both doesn't seem well suited due to the possible quirks with usize overflows.

Reply:
optimized that by hand